src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs

```diff
@@ -0,0 +1,298 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Composition;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Packaging;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Microsoft.VisualStudio.ComponentModelHost;
+using Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem;
+using Microsoft.VisualStudio.Shell.Interop;
+using NuGet.VisualStudio;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.Packaging
+{
+    [ExportWorkspaceServiceFactory(typeof(IPackageInstallerService)), Shared]
+    internal class PackageInstallerServiceFactory : IWorkspaceServiceFactory
+    {
+        public IWorkspaceService CreateService(HostWorkspaceServices workspaceServices)
+        {
+            return new PackageInstallerService();
+        }
+    }
+
+    internal class PackageInstallerService : ForegroundThreadAffinitizedObject, IPackageInstallerService
+    {
+        private readonly object _gate = new object();
+
+        private VisualStudioWorkspaceImpl _workspace;
+        private IVsPackageInstallerServices _packageInstallerServices;
+        private IVsPackageInstaller _packageInstaller;
+
+        private CancellationTokenSource _tokenSource = new CancellationTokenSource();
+        private bool _solutionChanged;
+        private HashSet<ProjectId> _changedProjects = new HashSet<ProjectId>();
+
+        private readonly ConcurrentDictionary<ProjectId, Dictionary<string, string>> _projectToInstalledPackageAndVersion =
+            new ConcurrentDictionary<ProjectId, Dictionary<string, string>>();
+
+        public bool TryInstallPackage(Workspace workspace, ProjectId projectId, string packageName, string versionOpt)
+        {
+            this.AssertIsForeground();
+
+            if (workspace == _workspace && _workspace != null && _packageInstallerServices != null)
+            {
+                var dte = _workspace.GetVsService<SDTE, EnvDTE.DTE>();
+                var dteProject = _workspace.TryGetDTEProject(projectId);
+                if (dteProject != null)
+                {
+                    try
+                    {
+                        if (!_packageInstallerServices.IsPackageInstalled(dteProject, packageName))
+                        {
+                            dte.StatusBar.Text = string.Format(ServicesVSResources.Installing_package_0, packageName);
+                            _packageInstaller.InstallPackage(source: null, project: dteProject, packageId: packageName, version: versionOpt, ignoreDependencies: false);
```

Comment:
What does it mean for `source` to be null? Does it have special meaning for NuGet.org? Does it mean "search NuGet.org no matter what" (in which case this will work, but it'd be _super_ weird) or does it mean "search all of my package sources", which will then fail if NuGet.org isn't one of your package sources?

If it's the latter, then can/should we prevent the suggestion of packages from Elfie when we know they don't have NuGet.org as a package source? Or should we suggest them anyway, just in case their package sources do contain it (which _might_ be likely if it's a very popular package)?